File: script/restart.py
#!/usr/bin/env python

import os
import platform
import subprocess
import sys
import time

def main():
  pid         = int(sys.argv[1])
  timeout     = int(sys.argv[2])
  system_os   = platform.system()
  startup_cmd = [os.path.abspath(arg) for arg in sys.argv[3:]]

  if system_os == 'Darwin' or system_os == 'Linux':
    if wait_for_exit_unix(pid, timeout):
      print(startup_cmd)
      subprocess.Popen(startup_cmd, close_fds=True)
      return 0
    else:
      print('Timeout exceeded.')
      return 1
  elif system_os == 'Windows':
    pass
  else:
    return 1

def wait_for_exit_unix(pid, timeout):
  wait_time = timeout
  while(wait_time > 0):
    try:
      os.kill(pid, 0)
    except OSError:
      return True
    time.sleep(1)
    wait_time -= 1000

  return False

def wait_for_exit_windows(timeout):
  wait_time = timeout
  while(wait_time > 0):
    time.sleep(1)
    wait_time -= 1000

  return False

if __name__ == '__main__':
  sys.exit(main())

File: script/restart.js
const childProcess  = require('child_process');
const path          = require('path');

const pid          = process.argv[2];
var timeout        = process.argv[3];
const systemOs     = process.platform;
var startupCmd     = '';
const waitForExit  = function(pid, timeout, cb) {
  waitTime = timeout
  var checkPid = setInterval(function() {
    if (waitTime > 0) {
      try {
        process.kill(pid, 0);
      } catch(e) {
        clearInterval(checkPid);
        cb(true);
      }
    } else {
      clearInterval(checkPid);
      cb(false);
    }

    waitTime -= 1;
  }, timeout * 1000);
};

for (var i = 4; i < process.argv.length; i++) {
  startupCmd = startupCmd + ' ' + (path.resolve(process.argv[i]));
}

if (timeout < 0)
  timeout = 10;
else if (timeout > 300)
  timeout = 300;

waitForExit(pid, timeout, function(success) {
  if (success) {
    childProcess.exec(startupCmd, function(err) {
      if (err) process.exit(1);
      process.exit();
    });
  } else {
    // timeout exceeded
    process.exit(1);
  }
});
